src/client/protocol.c

```diff
@@ -0,0 +1,832 @@
+#include <inttypes.h>
+#include <poll.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "../lib/assert.h"
+
+#include "protocol.h"
+#include "../message.h"
+#include "../protocol.h"
+#include "../request.h"
+#include "../response.h"
+#include "../tracing.h"
+#include "../tuple.h"
+
+/* Read data from fd into buf until one of the following occurs:
+ *
+ * - The full count n of bytes is read.
+ * - A read returns 0 (EOF).
+ * - Polling the fd in preparation for a read times out.
+ *
+ * On error, -1 is returned. Otherwise the return value is the count
+ * of bytes read. This may be less than n if either EOF happened or
+ * polling timed out. */
+static ssize_t doRead(int fd, void *buf, size_t n, int timeout_millis)
```

Comment:
Maybe we can pass the client struct here instead and do some accounting on the timeout and save e.g. `timeout_left` in the client struct after measuring the time spent in `poll`?